File: matrix/matrix_transposition.js
/**
 * Transposes a matrix
 * @param {number[][]} matrix Input matrix
 * @returns {number[][]} Transposed matrix
 */
function transpose(matrix) {
    if (!matrix.length) return [];

    const rows = matrix.length;
    const cols = matrix[0].length;

    const result = new Array(cols).fill(0).map(() => new Array(rows).fill(0));

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            result[j][i] = matrix[i][j];
        }
    }

    return result;
}

function test() {
    const matrix = [[1, 2], [3, 4]];
    const expected = [[1, 3], [2, 4]];
    const result = transpose(matrix);
    console.assert(JSON.stringify(result) === JSON.stringify(expected), "Test 1 Failed");

    const matrix2 = [[1, 2, 3], [4, 5, 6]];
    const expected2 = [[1, 4], [2, 5], [3, 6]];
    const result2 = transpose(matrix2);
    console.assert(JSON.stringify(result2) === JSON.stringify(expected2), "Test 2 Failed");

    console.log("All tests passed!");
}

test();

File: matrix/matrix_transposition.cpp
/**
 * @file
 * @brief Implementation of Matrix Transposition
 * @details
 * Transposes a matrix (swaps rows and columns).
 * Time Complexity: O(R * C)
 * Space Complexity: O(R * C) for result matrix
 */

#include <cassert>
#include <iostream>
#include <vector>

/**
 * @brief Transposes a matrix
 * @param matrix Input matrix
 * @return Transposed matrix
 */
std::vector<std::vector<int>>
transpose(const std::vector<std::vector<int>> &matrix) {
  if (matrix.empty())
    return {};

  size_t rows = matrix.size();
  size_t cols = matrix[0].size();

  std::vector<std::vector<int>> result(cols, std::vector<int>(rows));

  for (size_t i = 0; i < rows; ++i) {
    for (size_t j = 0; j < cols; ++j) {
      result[j][i] = matrix[i][j];
    }
  }

  return result;
}

/**
 * @brief Self-test implementations
 */
void test() {
  std::vector<std::vector<int>> matrix = {{1, 2}, {3, 4}};
  std::vector<std::vector<int>> expected = {{1, 3}, {2, 4}};
  assert(transpose(matrix) == expected);

  std::vector<std::vector<int>> matrix2 = {{1, 2, 3}, {4, 5, 6}};
  std::vector<std::vector<int>> expected2 = {{1, 4}, {2, 5}, {3, 6}};
  assert(transpose(matrix2) == expected2);

  std::cout << "All tests passed!" << std::endl;
}

int main() {
  test();
  return 0;
}
